pkg/backup/compaction_processor.go

```diff
@@ -0,0 +1,465 @@
+			log.Warningf(ctx, "failed to flush sink: %v", err)
+			logClose(ctx, sink, "SST sink")
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case entry, ok := <-entryCh:
+			if !ok {
+				return nil
+			}
+
+			sstIter, err := openSSTs(ctx, execCfg, entry, fileEncryption, spec.EndTime)
+			if err != nil {
+				return errors.Wrap(err, "opening SSTs")
+			}
+
+			if err := compactSpanEntry(ctx, sstIter, sink); err != nil {
+				return errors.Wrap(err, "compacting span entry")
+			}
+		}
+	}
+}
+
+func compactionChainFromSpec(
+	ctx context.Context,
+	execCfg *sql.ExecutorConfig,
+	spec execinfrapb.CompactBackupsSpec,
+	user username.SQLUsername,
+) (compactionChain, *jobspb.BackupEncryptionOptions, error) {
+	kmsEnv := backupencryption.MakeBackupKMSEnv(
+		execCfg.Settings,
+		&execCfg.ExternalIODirConfig,
+		execCfg.InternalDB,
+		user,
+	)
+	prevManifests, localityInfo, encryption, allIters, err := getBackupChain(
+		ctx, execCfg, user, spec.Destination, spec.Encryption, spec.EndTime, &kmsEnv,
+	)
+	if err != nil {
+		return compactionChain{}, nil, err
+	}
+	compactChain, err := newCompactionChain(
+		prevManifests, spec.StartTime, spec.EndTime, localityInfo, allIters,
+	)
+	if err != nil {
+		return compactionChain{}, nil, err
+	}
+	return compactChain, encryption, nil
+}
+
+func genIntersectingSpansForCompaction(
+	ctx context.Context,
+	execCfg *sql.ExecutorConfig,
+	spec execinfrapb.CompactBackupsSpec,
+	compactChain compactionChain,
+	intersectSpanCh chan execinfrapb.RestoreSpanEntry,
+) error {
+	backupLocalityMap, err := makeBackupLocalityMap(compactChain.compactedLocalityInfo, spec.User())
+	if err != nil {
+		return err
+	}
+
+	introducedSpanFrontier, err := createIntroducedSpanFrontier(
+		compactChain.backupChain, compactChain.lastBackup().EndTime,
+	)
+	if err != nil {
+		return err
+	}
+	defer introducedSpanFrontier.Release()
+
+	targetSize := targetRestoreSpanSize.Get(&execCfg.Settings.SV)
+	maxFiles := maxFileCount.Get(&execCfg.Settings.SV)
+
+	var fsc fileSpanComparator = &exclusiveEndKeyComparator{}
+	filter, err := makeSpanCoveringFilter(
+		spec.Spans,
+		[]jobspb.RestoreProgress_FrontierEntry{},
+		introducedSpanFrontier,
+		targetSize,
+		maxFiles,
+	)
+	if err != nil {
+		return err
+	}
+
+	genSpan := func(ctx context.Context, entryCh chan execinfrapb.RestoreSpanEntry) error {
+		defer close(entryCh)
+		return errors.Wrapf(generateAndSendImportSpans(
+			ctx,
+			spec.Spans,
+			compactChain.chainToCompact,
+			compactChain.compactedIterFactory,
+			backupLocalityMap,
+			filter,
+			fsc,
+			entryCh,
+		), "generate and send import spans")
+	}
+
+	filterIntersectingSpans := func(ctx context.Context, entryCh chan execinfrapb.RestoreSpanEntry) error {
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case entry, ok := <-entryCh:
+				if !ok {
+					return nil
+				}
+				for _, assignedSpan := range spec.AssignedSpans {
+					if assignedSpan.Overlaps(entry.Span) {
```

Comment:
i think we should use assignSpan.Contains() instead. 

Perhaps as a runtime assertion, we should also error if an entry overlaps with an assigned span but is _not_ contained by it. That would imply that 2 procs are going to work on the same restore span entry, which would lead to sadness.

Reply:
👍 I do like avoiding sadness